powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/v3/EciesEncryptionBehavior.java

```diff
@@ -0,0 +1,214 @@
+
+            // Get decryptor for the application
+            final EciesDecryptor decryptor = eciesFactory.getEciesDecryptorForApplication((ECPrivateKey) privateKey, applicationSecret);
+
+            // Initialize decryptor with ephemeral public key
+            byte[] ephemeralPublicKeyBytes = BaseEncoding.base64().decode(request.getEphemeralKey());
+            decryptor.initEnvelopeKey(ephemeralPublicKeyBytes);
+
+            // Extract envelope key and sharedInfo2 parameters to allow decryption on intermediate server
+            final EciesEnvelopeKey envelopeKey = decryptor.getEnvelopeKey();
+            GetEciesDecryptorResponse response = new GetEciesDecryptorResponse();
+            response.setSecretKey(BaseEncoding.base64().encode(envelopeKey.getSecretKey()));
+            response.setSharedInfo2(BaseEncoding.base64().encode(decryptor.getSharedInfo2()));
+            return response;
+        } catch (InvalidKeySpecException | EciesException e) {
+            logger.warn("Get Ecies decryptor parameters for application failed, reason: {}", e);
+            throw localizationProvider.buildExceptionForCode(ServiceError.DECRYPTION_FAILED);
+        }
+    }
+
+    /**
+     * Get ECIES decryptor parameters for activation scope.
+     *
+     * @param request Request to get ECIES decryptor parameters.
+     * @return ECIES decryptor parameters for activation scope.
+     * @throws GenericServiceException In case ECIES decryptor parameters could not be extracted.
+     */
+    private GetEciesDecryptorResponse getEciesDecryptorParametersForActivation(GetEciesDecryptorRequest request) throws GenericServiceException {
+        if (request.getApplicationKey() == null || request.getEphemeralKey() == null) {
+            throw localizationProvider.buildExceptionForCode(ServiceError.DECRYPTION_FAILED);
+        }
+
+        try {
+            // Lookup the activation
+            final ActivationRecordEntity activation = repositoryCatalogue.getActivationRepository().findActivation(request.getActivationId());
+            if (activation == null) {
+                throw localizationProvider.buildExceptionForCode(ServiceError.ACTIVATION_NOT_FOUND);
+            }
+
+            // Check if the activation is in correct state
+            if (!ActivationStatus.ACTIVE.equals(activation.getActivationStatus())) {
+                throw localizationProvider.buildExceptionForCode(ServiceError.ACTIVATION_INCORRECT_STATE);
+            }
+
+            // Lookup the application
+            final ApplicationVersionEntity applicationVersion = repositoryCatalogue.getApplicationVersionRepository().findByApplicationKey(request.getApplicationKey());
+            if (applicationVersion == null) {
+                throw localizationProvider.buildExceptionForCode(ServiceError.INVALID_APPLICATION);
+            }
+            final ApplicationEntity application = applicationVersion.getApplication();
+
```

Comment:
Here, we should probably add a check that `application.getId()` is equal to `activation.getApplicationId()`, to see that the provided application key is actually associated with the same application as the activation...

Generally speaking, we are sending application key here mainly to check if the given version is still supported. In principle, we have everything we need just from the activation ID. We are able to look up the application and master key pair. Is the `application` variable used, at all?

Reply:
The `application` variable is used when getting the master server key:

```
final MasterKeyPairEntity masterKeyPairEntity = repositoryCatalogue.getMasterKeyPairRepository().findFirstByApplicationIdOrderByTimestampCreatedDesc(application.getId());
```

I will add the application ID check.